lib/chef/resource/selinux_module.rb

```diff
@@ -0,0 +1,130 @@
+
+  		property :cookbook, String,
+  		          description: 'Cookbook to source from module source file from(if it is not located in the current cookbook). The default value is the current cookbook.',
+                desired_state: false
+
+  		property :base_dir, String,
+  		          default: '/etc/selinux/local',
+  		          description: 'Directory to create module source file in'
+
+  		action_class do
+  		  def selinux_module_filepath(type)
+  		    path = ::File.join(new_resource.base_dir, "#{new_resource.module_name}")
+  		    path.concat(".#{type}") if type
+  		  end
+
+  		  def list_installed_modules
+  		    shell_out!('semodule --list-modules').stdout.split("\n").map { |x| x.split(/\s/).first }
+  		  end
+  		end
+
+  		action :create do
+  		  directory new_resource.base_dir
+
+  		  if property_is_set?(:content)
+  		    file selinux_module_filepath('te') do
+  		      content new_resource.content
+
+  		      mode '0600'
+  		      owner 'root'
+  		      group 'root'
+
+  		      action :create
+
+  		      notifies :run, "execute[Compiling SELinux modules at '#{new_resource.base_dir}']", :immediately
+  		    end
+  		  else
+  		    cookbook_file selinux_module_filepath('te') do
+  		      cookbook new_resource.cookbook
+  		      source new_resource.source
+
+  		      mode '0600'
+  		      owner 'root'
+  		      group 'root'
+
+  		      action :create
+
+  		      notifies :run, "execute[Compiling SELinux modules at '#{new_resource.base_dir}']", :immediately
+  		    end
+  		  end
+
+  		  execute "Compiling SELinux modules at '#{new_resource.base_dir}'" do
+  		    cwd new_resource.base_dir
+  		    command "make -C #{new_resource.base_dir} -f /usr/share/selinux/devel/Makefile"
+  		    timeout 120
+  		    user 'root'
+
+  		    action :nothing
+
+  		    notifies :run, "execute[Install SELinux module '#{selinux_module_filepath('pp')}']", :immediately
+  		  end
+
+  		  raise "Compilation must have failed, no 'pp' file found at: '#{selinux_module_filepath('pp')}'" unless ::File.exist?(selinux_module_filepath('pp'))
+
+  		  execute "Install SELinux module '#{selinux_module_filepath('pp')}'" do
+  		    command "semodule --install '#{selinux_module_filepath('pp')}'"
+  		    action :nothing
+  		  end
+  		end
+
+  		action :delete do
+  		  %w(fc if pp te).each do |type|
+  		    next unless ::File.exist?(selinux_module_filepath(type))
+
+  		    file selinux_module_filepath(type) do
+  		      action :delete
+  		    end
+  		  end
+  		end
+
+  		action :install do
+  		  raise "Module must be compiled before it can be installed, no 'pp' file found at: '#{selinux_module_filepath('pp')}'" unless ::File.exist?(selinux_module_filepath('pp'))
+
+  		  unless list_installed_modules.include? new_resource.module_name
+  		    converge_by "Install SELinux module #{selinux_module_filepath('pp')}" do
+  		      shell_out!("semodule --install '#{selinux_module_filepath('pp')}'")
```

Comment:
one thing i'd suggest is using:

```
shell_out!("semodule", "--install", selinux_module_filepath('pp'))
```

that avoids starting up a shell to parse the command line, and feeds the arguments into argv[0..2] and as a result the quoting of the 2nd argument is no longer necessary.

Reply:
Updated for now few occurrences. (Not sure if we should be splitting ones with pipe operator)